File: day06/lanternfish.py
import sys

fish = [0] * 9
with open(sys.argv[1]) as f:
    for x in map(int, f.readline().split(',')):
        fish[x] += 1

for day in range(256):
    tmp = fish[0]
    fish = fish[1:]
    fish[6] += tmp
    fish.append(tmp)

    if day == 79:
        print('Part 1:', sum(fish))
    elif day == 255:
        print('Part 2:', sum(fish))

File: day06/lanternfish.cpp
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <vector>
#include <numeric>

using namespace std;

int main(int argc, char *argv[]) {
    // initialize the buckets
    vector<unsigned long long> fish(9, 0);

    // parse the input
    const char *fname = argv[1];
    FILE *fp;
    if ((fp = fopen(fname, "r")) == NULL) {
        perror(fname);
        exit(1);
    }
    int c;
    while ((c = fgetc(fp)) != EOF)
        if (isdigit(c))
            fish[c - '0']++;
    fclose(fp);

    // now run the generations;
    for (int day = 1; day <= 256; day++) {
        vector<unsigned long long> tmp(fish.begin()+1, fish.end());
        tmp[6] += fish[0];
        tmp.push_back(fish[0]);
        fish = tmp;

        if (day == 80)
            printf("Part 1: %llu\n", accumulate(fish.begin(), fish.end(), 0LL));
        else if (day == 256)
            printf("Part 2: %llu\n", accumulate(fish.begin(), fish.end(), 0LL));
    }
}
